### PR TITLE
New Dockerfile (fix #1459)

### LINKED ISSUE #1459: Docker Image with v3.15.0
for the release v3.15.0 there is no Docker image on [Docker Hub](https://hub.docker.com/r/stackexchange/dnscontrol/tags) yet. Would it be possible to create one?

### PR DESCRIPTION
**Pull request by fdcastel**
My first attempt to solve the Docker imbroglio. 😅 

Based on the great work from @jauderho 👍 

Fix https://github.com/StackExchange/dnscontrol/issues/1459

### CODE REVIEW

**Comment by tlimoncelli** [comment]
> Considerations:
> 
> * Based on v3.16.2 release

:+1:


> * Using golang 1.18 (it won't build wit 1.16)

:+1:

> * `dnscontrol version` ir returning `dnscontrol v3.16.2-dev`. I don't know if this is expected or if I need to pass any other parameters to build

Non-docker builds are done via `go run build/build.go` which embeds the version number (without `-dev`).

Tom

**Comment by fdcastel** [comment]
Rebased from master. Build ok.

File: Dockerfile [hunk 1]
```diff
@@ -1,3 +1,33 @@
...
+    && update-ca-certificates \
+    && go build -v -trimpath -buildmode=pie -ldflags="-s -w"
+
+# Validation check
+RUN cp dnscontrol /go/bin/dnscontrol
+RUN dnscontrol version
+
+# build convertzone
+RUN go build -v -trimpath -buildmode=pie -ldflags="-s -w" -o cmd/convertzone/convertzone cmd/convertzone/main.go
+RUN cp cmd/convertzone/convertzone /go/bin/convertzone
```
[the remark this answers is not in the document]
Comment (fdcastel) [reply Dockerfile]: Got it! Done!

**Comment by fdcastel** [comment]
@jauderho I believe I did all the changes you suggested. Please ping me if I missed something.

@tlimoncelli We still need someone versed in CircleCI to fix the current build system, right?

**Comment by jauderho** [comment]
LGTM

**Comment by tlimoncelli** [comment]
> @tlimoncelli We still need someone versed in CircleCI to fix the current build system, right?

I may have misspoke earlier.  Most of the build is on CircleCI now.  So, adding the docker stuff shouldn't be too difficult.  The person that did the last update is on vacation for 2 weeks, but I should be able to guess my way through it :-)

File: Dockerfile [hunk 2]
```diff
@@ -1,7 +1,30 @@
...
+    && update-ca-certificates \
+    && go build -v -trimpath -buildmode=pie -ldflags="-s -w -X main.SHA=${BUILD_VERSION}"
+
+# Validation check
+RUN cp dnscontrol /go/bin/dnscontrol
+RUN dnscontrol version
+
+# -----
+
+FROM alpine:3.16.0@sha256:686d8c9dfa6f3ccfc8230bc3178d23f84eeaf7e457f36f271ab1acc53015037c
```
Comment (aba-rechsteiner) [inline Dockerfile:23]: @fdcastel why not from scratch?

**Comment by aba-rechsteiner** [comment]
@fdcastel  thank you for your work. But on docker hub it doesn't have a new image yet, is this no longer pushed to docker hub?

File: Dockerfile (hunk 2 above)
Comment (fdcastel) [reply Dockerfile:23]: I don't know. To the best of my understanding `scratch` is an empty image more appropriate for base O.S. images (like alpine, debian, etc.)

I have no experience with `FROM scratch` images. Opinions?

**Comment by fdcastel** [comment]
@aba-rechsteiner: It is not ready, yet.

As @tlimoncelli [said](https://github.com/StackExchange/dnscontrol/pull/1537#issuecomment-1155934068) we are waiting for the CircleCI maintainer, now.

Unfortunately I don't have any experience with CircleCI to help, in this case. 😞

File: Dockerfile (hunk 2 above)
Comment (aba-rechsteiner) [reply Dockerfile:23]: the dnscontrol go binary shouldn't have any external dependencies and therefore doesn't need a base os in the docker container, therefore it is sufficient to use the scratch image.

Comment (fdcastel) [reply Dockerfile:23]: I see.  @jauderho Opinions?

P.S.: A brief search pointed me this:

https://ashishb.net/all/docker-be-careful-about-the-scratch-image/

Dnscontrol doesn't have any static dependencies, but we may need some external services (e.g. Cloudflare, via https) in runtime.

I cannot test it right now but will asap.

Comment (aba-rechsteiner) [reply Dockerfile:23]: @fdcastel yes the COPY line should be there in any case
`COPY --from=alpine /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/`

Comment (jauderho) [reply Dockerfile:23]: @aba-rechsteiner Have you actually tried to build this with scratch as the base image? It doesn't work for me which is why I settled on using alpine.

**Pull request merged**